compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala

```diff
@@ -1245,6 +1245,21 @@ class TreeUnpickler(reader: TastyReader,
               skipTree()
               readStats(ctx.owner, end,
                 (stats, ctx) => Block(stats, exprReader.readTerm()(using ctx)))
+            case ASSUMEINFO =>
+              val acc = scala.collection.mutable.ListBuffer.empty[AssumeInfo]
+              def rec(): List[AssumeInfo] =
+                acc += AssumeInfo(readSymRef(), readInt(), readType(), EmptyTree)
+                if nextByte == ASSUMEINFO then
+                  readByte() // ASSUMEINFO
+                  readEnd()
+                  rec()
+                else acc.toList
+              val assumeInfos = rec()
+              inContext(ctx.withGadt(ctx.gadt.withAssumeInfos(assumeInfos))) {
+               assumeInfos.foldRight(readTerm()) { case (AssumeInfo(sym, nestingLevel, info, _), body) =>
```

Comment:
I guess I don't have a canonical example that would fail without this.  I haven't tried (today) to take it out and test - but I can.  But my general desire is to have these known facts to always be present whenever there's a chance there's some type comparisons potentially happening.  As I understand it the design is that there's only a very light, simple amount of type composition that happens during unpickling (e.g. `List + Int = List[Int]`).  But does none of it require any type comparisons?

What's your concern here?  Performance?

Reply:
My concern is mostly whether we can replicate this logic in tasty-query. In tasty-query, the unpickling step is not allowed to look at any information coming from outside the current `.tasty` file. If dotty requires contextual type information like this to unpickle, there's a good chance that tasty-query would need it as well, and that would be a problem.

Reply:
I'd prefer not to enrich the context. First. it costs time. Second, I don't see a case where we would need it. Which means we should start without and see whether we are proven wrong. If we are proven wrong it might be that TreeUnpickler does more than we think it should do, which would require a fix there. Or we conclude that, indeed, adding assumed infos is necessary. But then in either case we have learned something. By contrast, adding stuff "just in case" tells us nothing.

Reply:
I like that logic, and I agree.

Sadly it breaks the following:

```
    tests/pos/gadt-GadtStlc.scala failed
    tests/pos/i14776.scala failed
    tests/pos/i14776-patmat.scala failed
    tests/pos/gadt-cast-if.scala failed
```

Because the trees from pickling don't have the same types as the original.

Reply:
Maybe that's a deal breaker then. I believe that TastyQuery (or any of the other tools) should not have to know about Gadt contexts. So, maybe we should abandon the effort and stick with casts.

Reply:
Maybe my problem is that GadtConstraint itself is too complicated. To start with, it sits in the wrong place of the type comparer system: it should be a constraint, but it is really a constraint handler (i.e. a mutable abstraction that encapsulates a constraint field). Then there is all this complication of creating type variables for a real constraint that sits behind it. For type inference, all of this is a bit awkward, and should really be redesigned before we go further. For after type inference, it should have no place whatsoever. 

So if assumeInfo should work, it should be decoupled from GadtConstraints

My recommendation now would be to stop all work on extending GadtConstraints until we have a better core design that we can actually understand.

Reply:
What we could do (maybe, eventually): 

 - Change the context so that instead of a GadtConstraint it just contains a map from symbols to infos. 
 - TypeComparer would consult that map and, if a symbol has an entry, try that entry for comparisons.
 - The entry _could be_ a type variable linked to a GadtConstraint, but that's the business of Typer. `Context` knows nothing about this.

Reply:
> * Change the context so that instead of a GadtConstraint it just contains a map from symbols to infos.
> * TypeComparer would consult that map and, if a symbol has an entry, try that entry for comparisons.

FWIW, this is also how I imagine tasty-query would deal with `AssumeInfo`, *and* with `opaque type` aliases. A map of symbols to infos for `AssumeInfo` handling, and a set of opaque type alias symbols that must be treated as translucent.

Reply:
That all sounds good to me.  I'll see what I can do and we can chat about it.